# Table Of Contents:
[1 - Introduction](#1---Introduction)
* [1.1 - Data Description](#1.1----Data-Description)
* [1.2 - Rename Columns](#1.2---Rename-Columns)
* [1.3 - Check Column](#1.3---Check-Column)

[2 - Handle Columns](#2----Handle-Columns)
* [2.1 - "Unnamed: 13" column](#2.1----"Unnamed:-13"-column)
* [2.2 - "Unnamed: 14" column](#2.2----"Unnamed:-14"-column)
* [2.3 - "Unnamed: 15" column](#2.3---"Unnamed:-15"-column)
* [2.4 - "Unnamed: 16" column](#2.4----"Unnamed:-16"-column)

[3 - Recheck Records](#3---Recheck-Records)
- [3.1 - Replace Values](#3.1---Replace-Values)
- [3.2 - Assign DataFrame](#3.2---Assign-DataFrame)
- [3.3 - Remove Unnecessary Columns](#3.3---Remove-Unnecessary-Columns)
- [3.4 - Convert Data Types](#3.4---Convert-Data-Types)

[4 - Data Quality Report](#4---Data-Quality-Report)

[5 - Drop Rows](#5---Drop-Rows)

[6 - Download](#6---Download)

# 1 - Introduction

- Let's look at the image below. Do you notice anything strange ?:
    - `Category` values include entries like "50 Years in the Making", "Learn", etc.
    - `Main Category` includes entries like "Be active", "Clothing company",etc
    -  `Deadline`, `currency`, and `several other columns` show **inconsistent formats** and **redundant columns** with **4 Unnamed columns**.
- The current dataset is not only dirty, but the structure is also compromised. The misallignment stems from **the comma** ',' inside `name` column during import.
- Let's clean it up, because **bad data kills good decisions**.

![Biểu đồ EDA](Picture/1.png) 

# 1.1 -  Data Description
- ID: Kickstarter project ID
- name: Name of project
- category: Category of project
- main_category: Main category of project
- goal: Fundraising goal
- pledged: Amount pledged
- state: State of project (successful, canceled, etc.)
- backers: Number of project backers

In [1]:
import pandas as pd
import numpy as np
import datetime as dt

In [2]:
kicks_2016 = pd.read_csv('ks-projects-201612.csv', encoding='latin1', low_memory=False, quotechar='"')

In [3]:
kicks_2016.columns

Index(['ID ', 'name ', 'category ', 'main_category ', 'currency ', 'deadline ',
       'goal ', 'launched ', 'pledged ', 'state ', 'backers ', 'country ',
       'usd pledged ', 'Unnamed: 13', 'Unnamed: 14', 'Unnamed: 15',
       'Unnamed: 16'],
      dtype='object')

As mentioned earlier, there are four redundant columns :`'Unnamed: 13'`, `'Unnamed: 14'`, `'Unnamed: 15`',`'Unnamed: 16'`. This likely happened because **some project names contain multiple commas**.

## 1.2 - Rename Columns

First of all, we can see a white space behind the name of each column. Let's remove unwanted line breaks first.

In [4]:
kicks_2016.rename({
    "ID ": "ID",
    "name ": "name",
    "main_category ":"main_category",
    "currency ": "currency",
    "deadline ": "deadline",
    "goal ":"goal",
    "launched ": "launched",
    "pledged ": "pledged",
    "state ": "state",
    "country ":"country",
    "usd pledged ": "usd pledged",
    "backers ":"backers",
    "category " :"category"
}, axis = 1, inplace=True)

## 1.3 - Check Column

In [5]:
kicks_2016['goal'].value_counts().tail(50)

goal
50022                  1
7410                   1
7425                   1
10035.35               1
2921                   1
64550                  1
20510                  1
704                    1
34760                  1
8118                   1
1167                   1
96                     1
11652                  1
6955                   1
16557                  1
49812                  1
24410                  1
13860                  1
12399                  1
18011                  1
2016-11-13 17:51:53    1
21040                  1
3120                   1
835000                 1
6977                   1
359000                 1
13560                  1
4395                   1
6396                   1
209731                 1
6127                   1
952                    1
448129                 1
1208                   1
56750                  1
2014-08-06 18:26:36    1
9077                   1
5214                   1
2010-09-03 02:27:00    1
9187                

The `goal` column is not supposed to contain data belongs to `deadline` column.

In [6]:
kicks_2016.loc[kicks_2016['goal'] == '2011-08-08 04:00:00']

ID          name                              category  \
301534  866525923  Womanthology   Massive All Female Comic Anthology!   

       main_category currency deadline                 goal launched  \
301534        Comics   Comics      USD  2011-08-08 04:00:00    25000   

                    pledged      state     backers country usd pledged  \
301534  2011-07-08 01:19:43  109301.56  successful    2001          US   

       Unnamed: 13 Unnamed: 14 Unnamed: 15  Unnamed: 16  
301534   109301.56         NaN         NaN          NaN

# 2 -  Handle Columns
- Step 1:To ensure we don't affect the original dataset, we'll first create a copy of the table
- Step 2: Define `condition`:
    - Use str.isnumeric() to flag values that are fully numeric → returns True for integers.
    - Additionally, any value that contains a decimal point '.' will also be considered True, as these represent float numbers.

In [7]:
s2 = kicks_2016.copy()
s2['condition1'] = s2['goal'].str.isnumeric() | s2['goal'].str.contains(r'\.')

In [8]:
s2['country'].unique()

array(['GB', 'US', 'CA', 'NO', 'AU', 'IT', 'DE', 'IE', 'ES', 'N,"0', 'MX',
       'SE', 'FR', 'NL', 'NZ', 'CH', 'AT', 'BE', 'DK', 'HK', 'LU', '14',
       '46', '10', '0', '42', 'SG', '35', '780', '122', '213', '2', '3',
       '126', '73', '19', '1', '142', 'failed', '140', '38', '24', '60',
       '37', '29', '167', '45', '23', '4', '85', '9', '248', '111', '18',
       '13', '28', '40', '54', '17', '20', '195', '25', '72', '22', '12',
       '158', '62', '307', '31', '107', '7', '110', '6', '69', '68', '84',
       '75', '119', '78', '150', '77', '83', '89', '16', '154', '92',
       '11', '116', '56', '136', '58', '27', '26', '582', '43', '5', '41',
       '30', '177', '36', '15', '49', '51', '9854', '168', '8', '57',
       '283', '61', '33', '74', '157', '405', '91', '174', '813', '66',
       '52', '114', '288', '109', '21', '32', '169', '50', '98', '483',
       '289', '88', '124', '159', '222', '65', '631', '59', '53', '117',
       '197', 'successful', '70', '55', '108', '63'

- Function(assign_value) - explaination:
    - df: dataframe
    - changed_col: The column to be updated
    - condition: The column that holds the boolean condition
    - updated_col: Column to get the new value from if condition is True
    - unchanged_col: Column to get the original value from if condition is False

In [9]:
def assign_value(df,changed_col,condition,updated_col,unchanged_col):
    df[changed_col] = np.where(df[condition] == False, df[updated_col], df[unchanged_col])
    return df.head()

![Trường Hợp 1](Picture/TH1.png) 

- We're going to apply a straightforward way: 
    - `name = current_name + category`
        -  the curren `category` will be replaced by the value in the next column
    - The remaining columns will also shift left accordingly

In [10]:
s2['name'] = np.where(s2["condition1"]==False, (s2['name'] + ',' + s2['category']), s2['name'])

assign_value(s2,'category','condition1','main_category','category') # category = main_category
assign_value(s2,'main_category','condition1','currency','main_category') #main_category = currency
assign_value(s2,'currency','condition1','deadline','currency') #currency = deadline
assign_value(s2,'deadline','condition1','goal','deadline') #deadline = goal
assign_value(s2,'goal','condition1','launched','goal') #goal = launched
assign_value(s2,'launched','condition1','pledged','launched') # launched = pledged
assign_value(s2,'pledged','condition1','state','pledged') #pledged = state
assign_value(s2,'state','condition1','backers','state') #state = backers
assign_value(s2,'backers','condition1','country', 'backers') # backers = country
s2['country'] = np.where(s2["country"].str.isnumeric()==True, s2['usd pledged'], s2["country"]) #country = usd pledged
assign_value(s2,'usd pledged','condition1','Unnamed: 13','usd pledged') # usd pledged = Unnamed: 13

ID                                               name  \
0  1000002330                    The Songs of Adelaide & Abullah   
1  1000004038                                     Where is Hank?   
2  1000007540  ToshiCapital Rekordz Needs Help to Complete Album   
3  1000011046  Community Film Project: The Art of Neighborhoo...   
4  1000014025                               Monarch Espresso Bar   

         category main_category currency             deadline   goal  \
0          Poetry    Publishing      GBP  2015-10-09 11:36:00   1000   
1  Narrative Film  Film & Video      USD  2013-02-26 00:20:50  45000   
2           Music         Music      USD  2012-04-16 04:24:11   5000   
3    Film & Video  Film & Video      USD  2015-08-29 01:00:00  19500   
4     Restaurants          Food      USD  2016-04-01 13:38:27  50000   

              launched pledged       state backers country usd pledged  \
0  2015-08-11 12:12:28       0      failed       0      GB           0   
1  2013-01-12 00:20:50     220      failed       3      US         220   
2  2012-03-17 03:24:11       1      failed       1      US           1   
3  2015-07-04 08:35:03    1283    canceled      14      US        1283   
4  2016-02-26 13:38:27   52375  successful     224      US       52375   

  Unnamed: 13 Unnamed: 14 Unnamed: 15  Unnamed: 16  condition1  
0         NaN         NaN         NaN          NaN        True  
1         NaN         NaN         NaN          NaN        True  
2         NaN         NaN         NaN          NaN        True  
3         NaN         NaN         NaN          NaN        True  
4         NaN         NaN         NaN          NaN        True

We've finished cleaning up the table in the main columns of the table. Let's move  on to address the redundant columns

## 2.1 -  "Unnamed: 13" column

Before, we're going to create a function, let's check the data in `Unnamed: 13` column.

In [11]:
s2['Unnamed: 13'].unique()

array([nan, '504.94765278', '3984', '525', '0', '3035', '4986.41798748',
       '590', '23924', '12306', '13675', '3.39594898', '4690', '4488.01',
       '687', '450', '11719', '126.9669968', '1495', '20', '14309.36',
       'US', '8542', '3363', '950', '1800.5', '352', '1320', '18', '2161',
       '331', '13446.90533288', '2630', '1847.3965377', '3711', '103',
       '8609.6', '312', '11044', '5600.65', '630', '1010', '178',
       '2006.0578219', '3671', '65', '5', '4451', '16', '2000',
       '566.57479307', '125', '9444.43338774', '1', '90', '765',
       '357.66549924', '1850', '166', '2107', '63', '10', '8.5104355',
       '2920.0614495', '306', '13', '321.629504', '7703', '650', '800',
       '2275.2478569', '8.26779725', '1056', '5940', '45', '7.9006665',
       '1827', '10890.45', '245', '416.29712868', '200', '44.7846084',
       '6', '797', '6126', '175', '12201', '3605.13', '1200', '325',
       '2632', '6507.05253517', '46', '708.0265863', '9430.8', '2308',
       '310', '

- We're going to create a **second condition** to identify rows with missing data:
    - If the value is null -> Condition = True , 
    - If the value is not null -> Condition => False

In [12]:
s2['condition2'] = s2['Unnamed: 13'].isnull()

- It seems like all the data is the same as the column `usd_pledged`. We don't need to do anything except drop this column.

In [13]:
s2.loc[s2['condition2']==False]

ID                                               name  \
1454    1008705746  Zephyra´s new full length, 'As The World Colla...   
1563    1009317190           French Cuisine, A Traditional Experience   
1794    1010871699  The Beginners Guide to being Unsuicidal, the o...   
1931    1011687764       Best OnLine Classifieds, Ever / No More Spam   
2420     101453314  Social Media Ruined My Life, A Short Film from...   
...            ...                                                ...   
321945   989007729  THROUGH MY EYES, MY LIFE IN THE MISSISSIPPI DELTA   
322162   990511774  Daniel Hresko's new CD is (almost) ready, so g...   
322204   990746749             Feet on the Ground, Head in the Clouds   
323138   996542939  'WANDER' - Apocalyptic Short Film, Post Produc...   
323495    99851211  Art/sustainability/spirituality, a bike quest ...   

              category main_category currency             deadline   goal  \
1454             Metal         Music      SEK  2016-02-02 00:56:46  15000   
1563         Cookbooks          Food      USD  2014-09-08 00:46:23  13730   
1794           Theater       Theater      USD  2011-12-31 23:25:46   5000   
1931               Web    Technology      USD  2014-09-20 19:56:10   6300   
2420            Shorts  Film & Video      USD  2013-03-14 20:11:57   3000   
...                ...           ...      ...                  ...    ...   
321945  Narrative Film  Film & Video      USD  2012-04-05 02:34:49   3500   
322162      Indie Rock         Music      USD  2011-09-06 05:59:00    400   
322204    Film & Video  Film & Video      USD  2013-04-17 16:00:31  35000   
323138          Shorts  Film & Video      GBP  2015-04-26 17:12:59   1500   
323495     Photography   Photography      USD  2013-03-05 21:00:00   2500   

                   launched pledged       state backers country  \
1454    2016-01-03 00:56:46    4262      failed      14      SE   
1563    2014-08-09 03:16:02    3984      failed      46      US   
1794    2011-11-21 23:25:46     525      failed      10      US   
1931    2014-08-21 19:56:10       0      failed       0      US   
2420    2013-02-25 21:11:57    3035  successful      42      US   
...                     ...     ...         ...     ...     ...   
321945  2012-03-06 02:34:49      10      failed       1      US   
322162  2011-08-06 15:36:39      61      failed       4      US   
322204  2013-03-18 15:00:31     179      failed      10      US   
323138  2015-02-25 17:12:59    2501  successful      50      GB   
323495  2013-02-05 08:09:32    2634  successful      34      US   

          usd pledged    Unnamed: 13 Unnamed: 14 Unnamed: 15  Unnamed: 16  \
1454     504.94765278   504.94765278         NaN         NaN          NaN   
1563             3984           3984         NaN         NaN          NaN   
1794              525            525         NaN         NaN          NaN   
1931                0              0         NaN         NaN          NaN   
2420             3035           3035         NaN         NaN          NaN   
...               ...            ...         ...         ...          ...   
321945             10             10         NaN         NaN          NaN   
322162             61             61         NaN         NaN          NaN   
322204            179            179         NaN         NaN          NaN   
323138  3861.84447014  3861.84447014         NaN         NaN          NaN   
323495           2634           2634         NaN         NaN          NaN   

        condition1  condition2  
1454         False       False  
1563         False       False  
1794         False       False  
1931         False       False  
2420         False       False  
...            ...         ...  
321945       False       False  
322162       False       False  
322204       False       False  
323138       False       False  
323495       False       False  

[625 rows x 19 columns]

### 2.2 -  "Unnamed: 14" column
- The condition for `Unnamed: 14` is the same as `Unnamed: 13`

In [14]:
s2['condition3'] = s2['Unnamed: 14'].isnull()

In [15]:
s2.loc[s2['condition3']==False]

ID                                           name  \
13795   1081139420            The Rolling Stones, BEGGARS BANQUET   
16482   1097082409                        Pixel art, personalized   
61447   1365419849         VIA, ME (V.isual I.nteractive A.rchive   
104120  1618382802      Druid Hill Park Passport: Discover, Enjoy   
112934  1670090635        EXCALIBRE:Sword of Truth, Path of Light   
145502  1864561559                          T-shirt, dress shirts   
225692   410977509        Handmade Artisan Soaps: Natural, Gentle   
243349   516522480   American Booty, USA-made products Series One   
264243   642928449               FRANCIS - The Pope, Buenos Aires   
269970   677103185                          SixSixSeven , Angels    
275186   708058934                           Be Wise, Wear Wisdom   
305503   890269949  Community Clothing. Make Clothes, Create Jobs   

                       category       main_category      currency  \
13795    50 Years in the Making                Rock         Music   
16482         Retro Gaming art.              Crafts        Crafts   
61447        M.ercury E.dition)            Software    Technology   
104120                    Learn          Be active!    Publishing   
112934       Kingdom of Heaven.          Nonfiction    Publishing   
145502                    pants    clothing company       Apparel   
225692       Good for your skin              Crafts        Crafts   
243349                  Spirits            Painting           Art   
264243                   soccer   tango and culture  Film & Video   
269970                  Demons            Religion       Esoteric   
275186            Divine Wisdom             Apparel       Fashion   
305503            Restore Pride             Fashion       Fashion   

              deadline                 goal             launched  \
13795              USD  2011-08-12 01:17:48                 4625   
16482              USD  2016-07-28 21:36:59                  500   
61447              USD  2016-03-19 20:40:23                10000   
104120      Publishing                  USD  2012-07-28 01:30:00   
112934             USD  2016-04-01 08:00:00                24000   
145502         Fashion                  USD  2015-08-02 18:23:23   
225692             USD  2015-10-08 22:08:39                  200   
243349             USD  2013-02-02 03:45:14                 1776   
264243    Film & Video                  USD  2014-03-02 23:34:00   
269970  Graphic Novels               Comics                  USD   
275186             USD  2015-02-27 13:53:56                10000   
305503             GBP  2016-03-16 16:24:27                75000   

                    pledged                state              backers  \
13795   2011-06-04 01:17:48                   20               failed   
16482   2016-06-28 21:36:59                    0               failed   
61447   2016-02-18 21:40:23                   26               failed   
104120                 9500  2012-06-06 23:54:14                 9854   
112934  2016-02-22 21:57:30                    1               failed   
145502                20000  2015-07-03 18:23:23                    0   
225692  2015-09-08 22:08:39                  435           successful   
243349  2013-01-03 03:45:14                  401               failed   
264243                75000  2014-02-05 02:06:20                    0   
269970  2015-10-10 01:00:00                  750  2015-09-10 18:15:45   
275186  2015-01-28 13:53:56                    1               failed   
305503  2016-02-15 17:24:27              88619.2           successful   

                    country usd pledged Unnamed: 13       Unnamed: 14  \
13795                failed          US          US                20   
16482                failed          US          US                 0   
61447                failed          US          US                26   
104120           successful         208         208                US   
112934              

The function for unnamed_14 is the same as when we handle in the columns. Only `Backers` and `country` columns require retrieving values from `Unnamed: 13 & 14`.For `usd pledged`, there is no corresponding data. So we fill the entry with **np.nan**

In [16]:
 def assign_value_unnamed_14(df,changed_col,condition3,updated_col,unchanged_col):
    df[changed_col] = np.where(df[condition3] == False , df[updated_col], df[unchanged_col])
    return df.head()

In [17]:
s2['name'] = np.where(s2['condition3']==False, s2['name'] + ',' + s2['category'],s2['name'])

assign_value_unnamed_14(s2,'category','condition3','main_category','category')
assign_value_unnamed_14(s2,'main_category','condition3','currency', 'main_category')
assign_value_unnamed_14(s2,'currency','condition3','deadline','currency')
assign_value_unnamed_14(s2,'deadline','condition3','goal','deadline')
assign_value_unnamed_14(s2,'goal','condition3','launched','goal')
assign_value_unnamed_14(s2,'launched','condition3','pledged','launched')
assign_value_unnamed_14(s2,'pledged','condition3','state','pledged')
assign_value_unnamed_14(s2,'state','condition3','backers','state')
assign_value_unnamed_14(s2,'backers','condition3','Unnamed: 14','backers')
assign_value_unnamed_14(s2,'country','condition3','Unnamed: 13','country')

#gán giá trị null cho cột usd_pledged
s2['usd pledged'] = np.where(s2['condition3']==False,np.nan ,s2['usd pledged'] )

In [18]:
s2.loc[(s2['condition3']==False) & (s2['Unnamed: 15'].isnull()==True)]

ID                                               name  \
13795   1081139420  The Rolling Stones, BEGGARS BANQUET, 50 Years ...   
16482   1097082409         Pixel art, personalized, Retro Gaming art.   
61447   1365419849  VIA, ME (V.isual I.nteractive A.rchive, M.ercu...   
112934  1670090635  EXCALIBRE:Sword of Truth, Path of Light, Kingd...   
225692   410977509  Handmade Artisan Soaps: Natural, Gentle, Good ...   
243349   516522480  American Booty, USA-made products Series One, ...   
275186   708058934                Be Wise, Wear Wisdom, Divine Wisdom   
305503   890269949  Community Clothing. Make Clothes, Create Jobs,...   

          category main_category currency             deadline   goal  \
13795         Rock         Music      USD  2011-08-12 01:17:48   4625   
16482       Crafts        Crafts      USD  2016-07-28 21:36:59    500   
61447     Software    Technology      USD  2016-03-19 20:40:23  10000   
112934  Nonfiction    Publishing      USD  2016-04-01 08:00:00  24000   
225692      Crafts        Crafts      USD  2015-10-08 22:08:39    200   
243349    Painting           Art      USD  2013-02-02 03:45:14   1776   
275186     Apparel       Fashion      USD  2015-02-27 13:53:56  10000   
305503     Fashion       Fashion      GBP  2016-03-16 16:24:27  75000   

                   launched  pledged       state           backers country  \
13795   2011-06-04 01:17:48       20      failed                20      US   
16482   2016-06-28 21:36:59        0      failed                 0      US   
61447   2016-02-18 21:40:23       26      failed                26      US   
112934  2016-02-22 21:57:30        1      failed                 1      US   
225692  2015-09-08 22:08:39      435  successful               435      US   
243349  2013-01-03 03:45:14      401      failed               401      US   
275186  2015-01-28 13:53:56        1      failed                 1      US   
305503  2016-02-15 17:24:27  88619.2  successful  128534.587723664      GB   

       usd pledged Unnamed: 13       Unnamed: 14 Unnamed: 15  Unnamed: 16  \
13795          NaN          US                20         NaN          NaN   
16482          NaN          US                 0         NaN          NaN   
61447          NaN          US                26         NaN          NaN   
112934         NaN          US                 1         NaN          NaN   
225692         NaN          US               435         NaN          NaN   
243349         NaN          US               401         NaN          NaN   
275186         NaN          US                 1         NaN          NaN   
305503         NaN          GB  128534.587723664         NaN          NaN   

        condition1  condition2  condition3  
13795        False       False       False  
16482        False       False       False  
61447        False       False       False  
112934       False       False       False  
225692       False       False       False  
243349       False       False       False  
275186       False       False       False  
305503       False       False       False

### 2.3 - "Unnamed: 15" column

In [19]:
s2['condition4'] = s2['Unnamed: 15'].isnull()

you'll notice that the method is the same as previous cases -  except the  Row Index **269970**.

In [20]:
s2.loc[s2['condition4']==False,"name":"condition4"]

name            category  \
104120  Druid Hill Park Passport: Discover, Enjoy, Learn          Be active!   
145502                      T-shirt, dress shirts, pants    clothing company   
264243          FRANCIS - The Pope, Buenos Aires, soccer   tango and culture   
269970                    SixSixSeven , Angels , Demons            Religion    

       main_category        currency deadline                 goal  \
104120    Publishing      Publishing      USD  2012-07-28 01:30:00   
145502       Apparel         Fashion      USD  2015-08-02 18:23:23   
264243  Film & Video    Film & Video      USD  2014-03-02 23:34:00   
269970      Esoteric  Graphic Novels   Comics                  USD   

                   launched              pledged                state backers  \
104120                 9500  2012-06-06 23:54:14                 9854      US   
145502                20000  2015-07-03 18:23:23                    0      US   
264243                75000  2014-02-05 02:06:20                    0      US   
269970  2015-10-10 01:00:00                  750  2015-09-10 18:15:45       1   

       country usd pledged Unnamed: 13 Unnamed: 14 Unnamed: 15  Unnamed: 16  \
104120     208         NaN         208          US        9854          NaN   
145502       0         NaN           0          US           0          NaN   
264243       0         NaN           0          US           0          NaN   
269970  failed         NaN      failed           1          US         25.0   

        condition1  condition2  condition3  condition4  
104120       False       False       False       False  
145502       False       False       False       False  
264243       False       False       False       False  
269970       False       False       False       False

In [21]:
def assign_value_unnamed_15(df,changed_col,condition4,updated_col,unchanged_col):
    df[changed_col] = np.where(df[condition4] == False, df[updated_col], df[unchanged_col])
    return df.head()

To avoid the function about the column `state` can affect delete the value 1 from the column backers to the index **269970**. 
We say that except the ID: !=677103185 => to not put the `underfined`.

In [22]:
s2['name'] = np.where(s2['condition4']==False,s2['name'] + ',' + s2['category'],s2['name'])

assign_value_unnamed_15(s2,'category','condition4','main_category','category')
assign_value_unnamed_15(s2,'main_category','condition4','currency','main_category')
assign_value_unnamed_15(s2,'currency','condition4','deadline','currency')
assign_value_unnamed_15(s2,'deadline', 'condition4', 'goal','deadline')
assign_value_unnamed_15(s2,'goal', 'condition4', 'launched','goal')
assign_value_unnamed_15(s2,'launched', 'condition4', 'pledged', 'launched')
assign_value_unnamed_15(s2,'pledged', 'condition4', 'state', 'pledged')

# undefined
s2['state'] = np.where((s2['condition4']==False) & (s2['ID'] != 677103185),'undefined', s2['state'])

assign_value_unnamed_15(s2, 'backers', 'condition4', 'country', 'backers')
assign_value_unnamed_15(s2, 'country', 'condition4', 'Unnamed: 14', 'country')
assign_value_unnamed_15(s2,'usd pledged', 'condition4', 'Unnamed: 15', 'usd pledged')

ID                                               name  \
0  1000002330                    The Songs of Adelaide & Abullah   
1  1000004038                                     Where is Hank?   
2  1000007540  ToshiCapital Rekordz Needs Help to Complete Album   
3  1000011046  Community Film Project: The Art of Neighborhoo...   
4  1000014025                               Monarch Espresso Bar   

         category main_category currency             deadline   goal  \
0          Poetry    Publishing      GBP  2015-10-09 11:36:00   1000   
1  Narrative Film  Film & Video      USD  2013-02-26 00:20:50  45000   
2           Music         Music      USD  2012-04-16 04:24:11   5000   
3    Film & Video  Film & Video      USD  2015-08-29 01:00:00  19500   
4     Restaurants          Food      USD  2016-04-01 13:38:27  50000   

              launched pledged       state  ... country usd pledged  \
0  2015-08-11 12:12:28       0      failed  ...      GB           0   
1  2013-01-12 00:20:50     220      failed  ...      US         220   
2  2012-03-17 03:24:11       1      failed  ...      US           1   
3  2015-07-04 08:35:03    1283    canceled  ...      US        1283   
4  2016-02-26 13:38:27   52375  successful  ...      US       52375   

  Unnamed: 13 Unnamed: 14 Unnamed: 15 Unnamed: 16  condition1  condition2  \
0         NaN         NaN         NaN         NaN        True        True   
1         NaN         NaN         NaN         NaN        True        True   
2         NaN         NaN         NaN         NaN        True        True   
3         NaN         NaN         NaN         NaN        True        True   
4         NaN         NaN         NaN         NaN        True        True   

   condition3  condition4  
0        True        True  
1        True        True  
2        True        True  
3        True        True  
4        True        True  

[5 rows x 21 columns]

In [23]:
s2.loc[s2['condition4']==False,"name":"condition4"]

name      category  \
104120  Druid Hill Park Passport: Discover, Enjoy, Lea...    Publishing   
145502     T-shirt, dress shirts, pants, clothing company       Apparel   
264243  FRANCIS - The Pope, Buenos Aires, soccer, tang...  Film & Video   
269970          SixSixSeven , Angels , Demons , Religion       Esoteric   

         main_category currency             deadline                 goal  \
104120      Publishing      USD  2012-07-28 01:30:00                 9500   
145502         Fashion      USD  2015-08-02 18:23:23                20000   
264243    Film & Video      USD  2014-03-02 23:34:00                75000   
269970  Graphic Novels   Comics                  USD  2015-10-10 01:00:00   

                   launched              pledged                state backers  \
104120  2012-06-06 23:54:14                 9854            undefined     208   
145502  2015-07-03 18:23:23                    0            undefined       0   
264243  2014-02-05 02:06:20                    0            undefined       0   
269970                  750  2015-09-10 18:15:45  2015-09-10 18:15:45  failed   

       country usd pledged Unnamed: 13 Unnamed: 14 Unnamed: 15  Unnamed: 16  \
104120      US        9854         208          US        9854          NaN   
145502      US           0           0          US           0          NaN   
264243      US           0           0          US           0          NaN   
269970       1          US      failed           1          US         25.0   

        condition1  condition2  condition3  condition4  
104120       False       False       False       False  
145502       False       False       False       False  
264243       False       False       False       False  
269970       False       False       False       False

### 2.4 -  "Unnamed: 16" column

In [24]:
s2['condition5'] = s2['Unnamed: 16'].isnull()

In [25]:
s2.loc[s2['Unnamed: 16'].isnull()==False,'main_category':'condition5'] 

main_category currency deadline                 goal launched  \
269970  Graphic Novels   Comics      USD  2015-10-10 01:00:00      750   

                    pledged                state backers country usd pledged  \
269970  2015-09-10 18:15:45  2015-09-10 18:15:45  failed       1          US   

       Unnamed: 13 Unnamed: 14 Unnamed: 15  Unnamed: 16  condition1  \
269970      failed           1          US         25.0       False   

        condition2  condition3  condition4  condition5  
269970       False       False       False       False

In [26]:
def assign_value_unnamed_16(df,changed_col,condition5,updated_col,unchanged_col):
    df[changed_col] = np.where(df[condition5] == False, df[updated_col], df[unchanged_col])
    return df.head()

In [27]:
s2['name'] = np.where(s2['condition5']==False,s2['name'] + ',' + s2['category'],s2['name'])

assign_value_unnamed_16(s2,'category','condition5','main_category','category')
assign_value_unnamed_16(s2,'main_category','condition5','currency','main_category')
assign_value_unnamed_16(s2,'currency','condition5','deadline','currency')
assign_value_unnamed_16(s2,'deadline', 'condition5', 'goal','deadline')
assign_value_unnamed_16(s2,'goal', 'condition5', 'launched','goal')
assign_value_unnamed_16(s2,'launched', 'condition5', 'state','launched')
assign_value_unnamed_16(s2,'pledged', 'condition5', 'country', 'pledged')
assign_value_unnamed_16(s2, 'country', 'condition5', 'Unnamed: 15', 'country')
assign_value_unnamed_16(s2,'state', 'condition5', 'Unnamed: 13', 'state')
assign_value_unnamed_16(s2, 'backers', 'condition5', 'Unnamed: 14', 'backers')
assign_value_unnamed_16(s2,'usd pledged', 'condition5', 'Unnamed: 16', 'usd pledged')

ID                                               name  \
0  1000002330                    The Songs of Adelaide & Abullah   
1  1000004038                                     Where is Hank?   
2  1000007540  ToshiCapital Rekordz Needs Help to Complete Album   
3  1000011046  Community Film Project: The Art of Neighborhoo...   
4  1000014025                               Monarch Espresso Bar   

         category main_category currency             deadline   goal  \
0          Poetry    Publishing      GBP  2015-10-09 11:36:00   1000   
1  Narrative Film  Film & Video      USD  2013-02-26 00:20:50  45000   
2           Music         Music      USD  2012-04-16 04:24:11   5000   
3    Film & Video  Film & Video      USD  2015-08-29 01:00:00  19500   
4     Restaurants          Food      USD  2016-04-01 13:38:27  50000   

              launched pledged       state  ... usd pledged Unnamed: 13  \
0  2015-08-11 12:12:28       0      failed  ...           0         NaN   
1  2013-01-12 00:20:50     220      failed  ...         220         NaN   
2  2012-03-17 03:24:11       1      failed  ...           1         NaN   
3  2015-07-04 08:35:03    1283    canceled  ...        1283         NaN   
4  2016-02-26 13:38:27   52375  successful  ...       52375         NaN   

  Unnamed: 14 Unnamed: 15 Unnamed: 16 condition1  condition2  condition3  \
0         NaN         NaN         NaN       True        True        True   
1         NaN         NaN         NaN       True        True        True   
2         NaN         NaN         NaN       True        True        True   
3         NaN         NaN         NaN       True        True        True   
4         NaN         NaN         NaN       True        True        True   

   condition4  condition5  
0        True        True  
1        True        True  
2        True        True  
3        True        True  
4        True        True  

[5 rows x 22 columns]

In [28]:
s2.loc[s2['Unnamed: 16'].isnull()==False,'name':'condition5'] 

name        category  \
269970  SixSixSeven , Angels , Demons , Religion , Eso...  Graphic Novels   

       main_category currency             deadline goal             launched  \
269970        Comics      USD  2015-10-10 01:00:00  750  2015-09-10 18:15:45   

       pledged   state backers  ... usd pledged Unnamed: 13 Unnamed: 14  \
269970       1  failed       1  ...        25.0      failed           1   

       Unnamed: 15 Unnamed: 16  condition1  condition2  condition3  \
269970          US        25.0       False       False       False   

        condition4  condition5  
269970       False       False  

[1 rows x 21 columns]

All values have now been cleaned and rearranged into their correct columns, restoring the dataset’s original structure.

# 3 - Recheck Records

In [29]:
s2.columns

Index(['ID', 'name', 'category', 'main_category', 'currency', 'deadline',
       'goal', 'launched', 'pledged', 'state', 'backers', 'country',
       'usd pledged', 'Unnamed: 13', 'Unnamed: 14', 'Unnamed: 15',
       'Unnamed: 16', 'condition1', 'condition2', 'condition3', 'condition4',
       'condition5'],
      dtype='object')

In [30]:
def unique_value(df, col_name):
    display(df[col_name].unique())

In [31]:
unique_value(s2,'country')

array(['GB', 'US', 'CA', 'NO', 'AU', 'IT', 'DE', 'IE', 'ES', 'N,"0', 'MX',
       'SE', 'FR', 'NL', 'NZ', 'CH', 'AT', 'BE', 'DK', 'HK', 'LU', 'SG'],
      dtype=object)

## 3.1 - Replace Values

In [32]:
s2['country'].replace({'N,"0': "NO"}, inplace=True)

C:\Users\admin\AppData\Local\Temp\ipykernel_17132\2853669558.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  s2['country'].replace({'N,"0': "NO"}, inplace=True)


## 3.2 - Assign DataFrame

In [33]:
kicks_2016 = s2

## 3.3 - Remove Unnecessary Columns

In [34]:
kicks_2016 = kicks_2016.drop(columns=['Unnamed: 13','Unnamed: 14','Unnamed: 15','Unnamed: 16',
                                      'condition1', 'condition2', 'condition3', 'condition4', 'condition5'])

## 3.4 - Convert Data Types

In [35]:
def convert_cols(df, col_name, dtypes):
    df[col_name] = df[col_name].astype(dtypes)
def convert_datetime(df,col_name):
    df[col_name] = pd.to_datetime(df[col_name])

In [36]:
convert_cols(kicks_2016,"goal",'Float64') #goal
convert_cols(kicks_2016, "pledged", float) #pledged
convert_cols(kicks_2016, "usd pledged", float) #usd_pledged
convert_datetime(kicks_2016,"deadline") #deadline
convert_datetime(kicks_2016,"launched") #launched

kicks_2016['backers'] = kicks_2016['backers'].round().astype('Float64').astype('Int64') #backers: float -> int

# 4 - Data Quality Report

In [37]:
def check_data_quality(df,column):
    quality_report = {
        'missing_value' : df.isnull().sum().to_dict(),
        'duplicate_value' : df[column].duplicated().sum(),
        'unique_value': len(df[column].unique()),
        'df_info': df.info()
    }
    return quality_report

In [38]:
quality_report = check_data_quality(kicks_2016,column='ID')
display(quality_report)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 323750 entries, 0 to 323749
Data columns (total 13 columns):
 #   Column         Non-Null Count   Dtype         
---  ------         --------------   -----         
 0   ID             323750 non-null  int64         
 1   name           323741 non-null  object        
 2   category       323750 non-null  object        
 3   main_category  323750 non-null  object        
 4   currency       323750 non-null  object        
 5   deadline       323750 non-null  datetime64[ns]
 6   goal           323750 non-null  Float64       
 7   launched       323750 non-null  datetime64[ns]
 8   pledged        323750 non-null  float64       
 9   state          323750 non-null  object        
 10  backers        323750 non-null  Int64         
 11  country        323750 non-null  object        
 12  usd pledged    319945 non-null  float64       
dtypes: Float64(1), Int64(1), datetime64[ns](2), float64(2), int64(1), object(6)
memory usage: 32.7+ MB


{'missing_value': {'ID': 0,
  'name': 9,
  'category': 0,
  'main_category': 0,
  'currency': 0,
  'deadline': 0,
  'goal': 0,
  'launched': 0,
  'pledged': 0,
  'state': 0,
  'backers': 0,
  'country': 0,
  'usd pledged': 3805},
 'duplicate_value': 0,
 'unique_value': 323750,
 'df_info': None}

# 5 - Drop Rows

In [39]:
# Name - null values
kicks_2016.loc[kicks_2016['name'].isnull()==True]

ID name            category main_category currency  \
36671   1218074363  NaN           Webseries  Film & Video      GBP   
41069    124438738  NaN  Interactive Design        Design      USD   
63544   1378236004  NaN             Apparel       Fashion      USD   
96753   1574873938  NaN                 Art           Art      USD   
142830  1848699072  NaN      Narrative Film  Film & Video      USD   
262901   634871725  NaN         Video Games         Games      GBP   
265251   648853978  NaN      Product Design        Design      USD   
269930   676846639  NaN              Design        Design      USD   
289847   796533179  NaN            Painting           Art      USD   

                  deadline      goal            launched   pledged  \
36671  2014-05-16 22:25:57    2000.0 2014-03-27 21:25:57      5.00   
41069  2015-03-25 20:51:48    9750.0 2015-02-23 21:51:48  10890.45   
63544  2016-05-28 00:00:00    1800.0 2016-04-29 21:04:40   2308.00   
96753  2014-04-03 17:52:09    7500.0 2014-03-04 17:52:09    100.00   
142830 2012-02-29 15:04:00  200000.0 2012-01-01 12:35:31    100.00   
262901 2013-01-06 23:00:00    2000.0 2012-12-19 23:57:48    196.00   
265251 2016-07-18 05:01:47    2500.0 2016-06-18 05:01:47      0.00   
269930 2011-04-04 20:08:17   50000.0 2011-03-05 20:08:17   1665.00   
289847 2011-12-05 05:59:00   35000.0 2011-11-06 23:55:55    220.00   

             state  backers country   usd pledged  
36671       failed        1      GB      8.267797  
41069   successful      107      US  10890.450000  
63544   successful      119      US   2308.000000  
96753       failed        1      US    100.000000  
142830      failed        1      US    100.000000  
262901      failed       12      GB    317.728436  
265251   suspended        0      US      0.000000  
269930    canceled       10      US   1665.000000  
289847      failed        5      US    220.000000

In [40]:
kicks_2016.drop(kicks_2016.loc[kicks_2016['name'].isnull()==True].index, inplace=True)

# 6 - Download

In [41]:
kicks_2016.to_csv('kicks_2016.csv', index=False)